src/opnsense/mvc/app/models/OPNsense/Interfaces/Vip.php

```diff
@@ -41,26 +41,43 @@ class Vip extends BaseModel
     public function performValidation($validateFullModel = false)
     {
         $messages = parent::performValidation($validateFullModel);
-        $vips = [];
+
+        $unqiue_addrs = [];
         $carp_vhids = [];
+        $vips = [];
 
         // collect changed VIP entries
         $vip_fields = ['mode', 'subnet', 'subnet_bits', 'password', 'vhid', 'interface'];
         foreach ($this->getFlatNodes() as $key => $node) {
             $tagName = $node->getInternalXMLTagName();
             $parentNode = $node->getParentNode();
+
             if ($validateFullModel || $node->isFieldChanged()) {
                 if ($parentNode->getInternalXMLTagName() === 'vip' && in_array($tagName, $vip_fields)) {
-                    $parentKey = $parentNode->__reference;
-                    $vips[$parentKey] = $parentNode;
+                    $vips[$parentNode->__reference] = $parentNode;
                 }
             }
+
             $vhid_key = sprintf("%s_%s", $parentNode->interface, $parentNode->vhid);
+
             if ((string)$parentNode->mode == 'carp' && !isset($carp_vhids[$vhid_key])) {
                 $carp_vhids[$vhid_key] = $parentNode;
             }
         }
 
+        foreach ($this->getFlatNodes() as $key => $node) {
```

Comment:
@fichtner I think I would combine this with the `foreach ($this->getFlatNodes() as $key => $node) {` loop above, so we can reuse some of it and only need to iterate once. Other than that, this looks nice :)

Reply:
fair enough. I moved it out of the loop because at first I tried to fit it into `if ($validateFullModel || $node->isFieldChanged()) {` which wasn't the case.